x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/hash/Hash.java

```diff
@@ -0,0 +1,115 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.expression.function.scalar.hash;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.compute.ann.Evaluator;
+import org.elasticsearch.compute.ann.Fixed;
+import org.elasticsearch.compute.operator.EvalOperator;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.tree.NodeInfo;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.expression.function.FunctionInfo;
+import org.elasticsearch.xpack.esql.expression.function.Param;
+import org.elasticsearch.xpack.esql.expression.function.scalar.EsqlScalarFunction;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamInput;
+
+import java.io.IOException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.HexFormat;
+import java.util.List;
+
+public class Hash extends EsqlScalarFunction {
+
+    public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(Expression.class, "Hash", Hash::new);
+
+    private final Expression alg;
+    private final Expression input;
+
+    @FunctionInfo(returnType = "keyword", description = "Computes the hash of the input using the supplied algorithm.")
+    public Hash(
+        Source source,
+        @Param(name = "alg", type = { "keyword", "text" }, description = "Hash algorithm to use.") Expression alg,
+        @Param(name = "input", type = { "keyword", "text" }, description = "Input to hash.") Expression input
+    ) {
+        super(source, List.of(alg, input));
+        this.alg = alg;
+        this.input = input;
+    }
+
+    private Hash(StreamInput in) throws IOException {
+        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class), in.readNamedWriteable(Expression.class));
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(alg);
+        out.writeNamedWriteable(input);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataType.KEYWORD;
+    }
+
+    @Override
+    public boolean foldable() {
+        return alg.foldable() && input.foldable();
+    }
+
+    @Evaluator(warnExceptions = NoSuchAlgorithmException.class)
+    static BytesRef process(BytesRef alg, BytesRef input) throws NoSuchAlgorithmException {
+        return hash(MessageDigest.getInstance(alg.utf8ToString()), input);
```

Comment:
Validate the algorithm at planning time. Use `Validatable` interface to verify the algorithm expression can be folded to a string and check the algorithm exist:

```
... implements Validatable {
    public void validate(Failures failures) {
           try { MessageDigest.getInstance(alg) 
           } catch (Exception ex) {
               failures.add("Invalid hashing algorithm [{}}", alg);
           }
    }
}
```

By doing the validation here, you can save the algorithm directly as a string instead as an expression.
Separately, you can validate the well known algorithms and use `org.es.common.MessageDigests` instead which has MD5 and SHA_1 variants.

Update: My comment is similar to Nik's

Reply:
If I understand correctly that is executed during the planning.
As a result we could only have constant `alg` inputs and will be unable to evaluate hash with algorithm read from the field or a variable. I wonder if we want to be able to dynamically resolve alg.

Reply:
If it *is* a constant then I think we should do what Costin says. I *think* the Validatable interface brings the failure much further forwards at planning time which is good. And it's a standard we'd like to stick to more closely, which is good. But for non-constant algorithms I'm fine supporting non-constant algorithms just as you have it.